src/lib/error/bad-data-error.ts

```diff
@@ -1,23 +1,90 @@
-class BadDataError extends Error {
-    constructor(message: string) {
-        super();
-        Error.captureStackTrace(this, this.constructor);
+import { ErrorObject } from 'ajv';
+import { getPropFromString } from '../util/get-prop-from-string';
+import { ApiErrorSchema, UnleashError } from './api-error';
 
-        this.name = this.constructor.name;
-        this.message = message;
+type ValidationErrorDescription = {
+    description: string;
+    message: string;
+    path?: string;
+};
+class BadDataError extends UnleashError {
+    details: ValidationErrorDescription[];
+
+    constructor(
+        message: string,
+        errors?: [ValidationErrorDescription, ...ValidationErrorDescription[]],
+    ) {
+        const topLevelMessage =
+            'Request validation failed: your request body contains invalid data' +
+            (errors
+                ? '. Refer to the `details` list to see what happened.'
+                : `: ${message}`);
+        super(topLevelMessage);
+
+        this.details = errors ?? [{ message: message, description: message }];
     }
 
-    toJSON(): object {
+    toJSON(): ApiErrorSchema {
         return {
-            isJoi: true,
-            name: this.constructor.name,
-            details: [
-                {
-                    message: this.message,
-                },
-            ],
+            ...super.toJSON(),
+            details: this.details,
         };
     }
 }
 
 export default BadDataError;
+
+export const fromOpenApiValidationError =
+    (requestBody: object) =>
```

Comment:
Depends on what you mean by "at this stage"? This is a function, so we're saying that we want an object, right?

Reply:
is it possible to put our system in a state where requestBody is not an object?

Reply:
When the req.body arrives over the wire its type is `unknown` until we validate/parse it into an object. My question is if we always call this function after validate/parse was done somewhere else in the code.

Reply:
Oh, right. TypeScript should prevent us from doing that, right (unknown isn't an object)? But yeah, this is only called if you get OpenAPI validation errors because something is wrong with your request. Shouldn't happen, but I'll verify that it doesn't break with an empty body. God catche

Reply:
Can confirm: if you try and post something without a body (and you mark it as json), it gets parsed into an empty object (presumably by express).